keps/sig-node/2043-pod-resource-concrete-assigments/README.md

```diff
@@ -215,6 +215,8 @@ ContainerDevices could represent it as following:
     },
 ]
 ```
+The cpu_ids field of ContainerResources and AllocatableResourcesResponse contains only CPU of the pool for exclusive isolation, but not
+from the shared pool.
```

Comment:
Sorry - I'm not able to comment on the PRR lines, so adding comments here:

- L300 - clearly Kubelet depends on the feature gate
- L309 - you can always imagine not catching something in CI - just say that Kubelets fails to start :)
- L338 - IIUC you're going to GA now - has this been addressed?
- L339 - how are watch notifications generated - are they generated when some pod is changing, then send and discarded, or it works differently?

Reply:
L338 - I didn't see any commits on this
L339 - action is reported such as ADDED/DELETED

Reply:
In line 338 you have:
```
In 1.18, DDOSing the API can lead to resource exhaustion. It is planned to be addressed as part of G.A.
```

I'm asking if that was addressed or what are the plans to address it.

> L339 - action is reported such as ADDED/DELETED

That I understand. I'm asking how/when they are generate (based on what signals), are they send only to streams that are only opened at this point, are they discarded immadiately after that etc.

Reply:
L338 - IIUC you're going to GA now - Looking [at the code](https://github.com/kubernetes/kubernetes/blob/9caf675ed1a055ebce18cf34e8284028a729fd00/pkg/kubelet/server/server.go#L182-L191), no interceptors or configuration is set to prevent DDOS. Is there a specific recommendation for the implementation here?

Reply:
> 
> 
> In line 338 you have:
> 
> ```
> In 1.18, DDOSing the API can lead to resource exhaustion. It is planned to be addressed as part of G.A.
> ```
> 
> I'm asking if that was addressed or what are the plans to address it.
> 
> > L339 - action is reported such as ADDED/DELETED
> 
> That I understand. I'm asking how/when they are generate (based on what signals), are they send only to streams that are only opened at this point, are they discarded immadiately after that etc.

@fromanirh can describe.how watch endpoint workes in details. As I remember if message was sent regarding some action, yes it is discarded.

Reply:
sure, thanks @AlexeyPerevalov for pointing this out. I'll provide the requested informations ASAP (need to finish something before)

Reply:
Hey @wojtek-t. Regarding Watch, things are still a bit in flux *exactly* for the stability/scalability concerns you are mentioning.
We are adding an endpoint to an API marked stable, we don't want absolutely to DOS the kubelet.
The current implementation is guided by the following:
1. generate watch events in a new goroutine from the main POD sync loop (IOW, never block the main POD sync loop!)
2. dispatch each watch event once to all the watch clients connected to the podresources unix socket
3. if no client is active watch events are always generated but immediately discarded (never stored)
4. never keep any event history. If a client connects too late, it must fetch the current state using other APIs (e.g. List())
5. handle slow clients is still open question. Surely avoid unbounded queues. Still to be discussed (PR review? seems implementation decision) if force disconnect slow clients, just discard old events (ring buffer for events, overwrite old events) or anything else.
Lacking better place, the plan is to add this content/design notes inside code comment in the PR
Last and not but least: I'm not aware of middleware/common components that we can reuse *for this API*. I'd be happy to consume them, just point them out.

Reply:
Thanks @fromanirh for explanation.

I think this kind of description should be definitely added to the KEP.

 Re DDOSing - I don't think we have any DDOS protection. I'm fine with leaving as is, just adding it to potential failure modes.

Reply:
@wojtek-t since we need to edit the KEP anyway, I'll make sure to add the design notes we just discussed. Thanks.

Reply:
Doesn't look like Wojtek's comments have been resolved, we're getting really close to enhancements freeze. For L319, should there be separate metrics for each endpoint (method)?

Reply:
1. Regarding DDOS - we decided to leave it as is
2. Watch endpoint is postponed, I hope this PR will update current state https://github.com/kubernetes/enhancements/pull/2402, I'll rebase after it will be merged. Watch API description will be added when implementation and the design will be ready. I think no need to add some special description on Watch API in this PR
3. Metrics: this PR about changing behavior of one field, this field affects both List and GetAllocatableResource endpoint. But should we use pod_resources_endpoint_requests_total for all endpoints, I think yes and it worth mentioning in this PR https://github.com/kubernetes/enhancements/pull/2404

Reply:
Ok, I would think you need an error count too. And ideally separate metrics per method in the API. I think I mentioned that in a related PR.

Reply:
@johnbelamaric metrics separated on line 320 of README.md as well as *_errors